Fix/7358 stock api performance improvement

moon0326 opened the pull request:
Fixes #7358 

This PR attempts to improve `low in stock` products REST API endpoint performance.

It adds a new endpoint: `/wp-json/wc-analytics/products/low-in-stock?page=1&per_page=5`

Original query: [gist](https://gist.github.com/moon0326/683fa74d8c606aa60f2ce8ed3b0d07db)
New query: [gist](https://gist.github.com/moon0326/bb49d48b257a1edaa22172759d604c74)

EXPLAIN of the original query (based on a store with 4k products):
![Screen Shot 2021-07-17 at 10 15 59 PM](https://user-images.githubusercontent.com/4723145/126056605-1b3c525e-f556-4c5c-9e04-c31e7f3de781.jpg)

EXPLAIN of the new query (based on a store with 4k products):

![Screen Shot 2021-07-17 at 10 16 28 PM](https://user-images.githubusercontent.com/4723145/126056612-e0dac81d-431a-4c4b-aea9-a398bb09310f.jpg)

In a store with about 10k products, the original query exceeds 30 seconds and times out. The new query takes around 200ms on average. The total query execution time should be 200ms (avg) + 60~80ms (wc_product_meta_lookup query). The query execution can be less than 100ms if the site is using sitewide low stock threshold only.     

### Changes

**The query selects `wc_product_meta_lookup` first then join `wp_posts`**

- This was suggested by @louwie17. I first thought we would not gain much performance, but realized that a store with a large dataset most likely has a higher # of rows in `wp_posts` than `wc_product_meta_lookup`. In a store with 10k products, this change showed 3x performance gain. This is the primary reason why I'm adding a new endpoint since modifying the existing endpoint to use `wc_product_meta_lookup` is quite tricky or nearly impossible without changing almost all of it.

- MySQL always uses `wp_posts` first with the original query, but MariaDB uses `wp_wc_order_product_lookup` depending on the # of rows in each table. This is optimized by the query optimizer automatically. If the query optimizer uses `wp_wc_order_product_lookup` internally anyway, it's safe to select `wp_wc_order_product_lookup` first.

**does not join wp_postmeta if the store is using sitewide stock threshold only**

If a store does not have any custom stock threshold set on a product then we don't have to join `wp_postmeta` table. You can see the method [here](https://github.com/woocommerce/woocommerce-admin/pull/7377/files#diff-d5ee2ec2a16bb74f2d14349119970c8a1725570fe70d1721523df9d40b0ad109R169).

The count query is an extra query, but the execution time is less than 0.1 ms and it can make the main query 2x+ faster if the store is using sitewide low stock threshold only.

**last_order_date is queried separately**

As you see in the EXPLAIN output of the original query, joining `wp_wc_order_product_lookup` is quite expensive. The estimated # of rows to be examined is 10062. The `last_order_date` is now [queried](https://github.com/woocommerce/woocommerce-admin/pull/7377/files#diff-d5ee2ec2a16bb74f2d14349119970c8a1725570fe70d1721523df9d40b0ad109R111) separately.


The query takes between 60~80ms in a store with the following `wc_order_product_lookup` dataset. This store has 10k products.

![Screen Shot 2021-07-17 at 10 40 22 PMth_](https://user-images.githubusercontent.com/4723145/126057042-700d9ff8-6e33-4b8a-a382-60cd407f1562.jpg)

**order by post_date is removed and we now order by wc_product_meta_lookup only**

Since we now select `wp_wc_order_product_lookup` first, ordering by a field that is not in the first selected table triggers `Using temporary; Using filesort`. These two should be avoided whenever possible for the best performance. 

Since `post_date` never gets updated and product_id is the same as post_id, it's safe to use the id field alone in this case.

### Notes

- I noticed that getting product images is also an expensive process as we query each product in a loop. This should be avoided whenever possible, but I'm leaving it as it is for now. We might want to work on this in a separate PR.

### Detailed test instructions:

[4kproducts.sql.zip](https://github.com/woocommerce/woocommerce-admin/files/6835729/4kproducts.sql.zip)

1. Download the `4kproducts.zip`. This SQL dump includes 4297 products for testing purposes.
2. Start with a fresh installation and DB. Import the SQL dump file. 
3. Navigate to WooCommerce -> Status -> Tools and click [ Regenerate ] button on the `Product lookup tables` row.
4. Confirm you have 9k+ rows in `wp_wc_order_product_lookup` table.
5. Navigate to WooCommerce -> Home and hide the task list then refresh.
6. You should see the activity panel with Orders, Stock, and Reviews.
7. Open inspector -> Network -> choose XHR then search for "low_in_stock". The API endpoint should be `products/low-in-stock`

If you want to compare how the new API performs against the existing API, checkout `b341b62b1d4eb7dcc5f088aa77c530ac893b65ec` commit. This commit still makes low-in-stock APIs to the existing endpoint.

Commit:
Add a new endpoint to get low in stock products.

Commit:
Set images and attributes property

Commit:
WIP -- set last order date

Commit:
Do not join postmeta if store does not have any custom stock threshold

louwie17 commented:
@moon0326 nice work on this! This is testing well for me, I assume if this goes through you would remove the original `low_in_stock` conditionals within the `Products.php` file?
Also should we still index `variation_id` as part of this PR, or not, given we aren't using it in these queries anymore.
cc: @jeffstieler on this.

Commit:
Combine get_query() and get_query_for_sitewide_threshold() into one method and a few code clean ups

Commit:
Support status field

Commit:
Use new products/low-in-stock endpoint to count low in stock products.

Commit:
Use new products/low-in-stock to get low in stock products data.

Commit:
Remove low_in_stock support

moon0326 commented:
> @moon0326 nice work on this! This is testing well for me, I assume if this goes through you would remove the original `low_in_stock` conditionals within the `Products.php` file?
> Also should we still index `variation_id` as part of this PR, or not, given we aren't using it in these queries anymore.
> cc: @jeffstieler on this.

Yup! Removed `low_in_stock` conditions in https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/39c7ae62e6dafaf7999470fe2842070e57a8e97b

I don't think we need an index for `variation_id` for this PR, but we should go over other API endpoints if we can benefit from the index in a separate issue IMO.

I made a few more updates (https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/bc63d54df5b1a4061c85f3d34b737cfe5ad98e2c, https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/ef4c21bdc968f6ebb204b24d2d92cdcbaa933245). You no longer need to update `Products.php` for testing.

adrianduffell commented:
Hey @moon0326 , thanks so much for taking this issue.

The code LGTM and I think this is close 🎉  I think adding the dedicated endpoint was a good choice to allow this query to be standalone and fine-tuned for its purpose.

What do you think of adding the same cache header to this endpoint as in as PR #7364?

```
$response->header( 'Cache-Control', 'max-age=300' );
```

I don't think it’s needed to solve the performance problem, but might be a nice general improvement. I think this data is fetched on every homescreen load at present?

> Yup! Removed low_in_stock conditions in 39c7ae6

We might need to keep this for the time being and deprecate it. I'm not sure if the mobile client depends on it? I'll find out.

I'm going to do further testing today, comparing it to the old query’s results.

moon0326 commented:
> Hey @moon0326 , thanks so much for taking this issue.
> 
> The code LGTM and I think this is close 🎉 I think adding the dedicated endpoint was a good choice to allow this query to be standalone and fine-tuned for its purpose.
> 
> What do you think of adding the same cache header to this endpoint as in as PR #7364?
> 
> ```
> $response->header( 'Cache-Control', 'max-age=300' );
> ```
> 
> I don't think it’s needed to solve the performance problem, but might be a nice general improvement. I think this data is fetched on every homescreen load at present?
> 
> > Yup! Removed low_in_stock conditions in [39c7ae6](https://github.com/woocommerce/woocommerce-admin/commit/39c7ae62e6dafaf7999470fe2842070e57a8e97b)
> 
> We might need to keep this for the time being and deprecate it. I'm not sure if the mobile client depends on it? I'll find out.
> 
> I'm going to do further testing today, comparing it to the old query’s results.

Thank you!

That's a good idea. I'll add the header.

What do you think about adding the following code to the existing `Products.php`? This should provide the same performance while keeping backward compatibility for those who are already using the endpoint.

Products.php -> get_items();
```
if ($request->get_param('low_in_stock') === true) {
	return (new ProductsLowInStock)->get_items($request);
}
```

adrianduffell commented:
> What do you think about adding the following code to the existing `Products.php`?

👍 Nice, that looks like a great approach!

It could still be deprecated by adding a note to the parameter description. For example like what was done here:

https://github.com/woocommerce/woocommerce-admin/blob/38fbb4aedd0f21cf21ddaa6af93e48564fd35324/src/API/Reports/Orders/Stats/Controller.php#L502-L510

Commit:
Deprecate low_in_stock

Commit:
Use the new ProductsLowInStock when low_in_stock is set

Commit:
Set cache-control header for the new endpoint

moon0326 commented:
@adrianduffell updated changes in the following commits.

Added deprecated comment: https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/a97c8f536db54560642bfb5c44918486c6bd054a

Use the new controller when `low_in_stock` is set: https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/f203979dd3eadeedc206799418c1711e24689104

Set cache-control: https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/bd33e8c4aabea8bf40c81247ae2d980213deeeea

adrianduffell commented:
Awesome, thanks @moon0326 

It looks like the cache header broke a test. I'm investigating.

On a related note, we probably need new PHP tests for the endpoint. In theory it could be a copy/paste of what's there for the deprecated endpoint?

moon0326 commented:
> Awesome, thanks @moon0326
> 
> It looks like the cache header broke a test. I'm investigating.
> 
> On a related note, we probably need new PHP tests for the endpoint. In theory it could be a copy/paste of what's there for the deprecated endpoint?

Thank you! 

Yeah, I think we can extract out low stock related methods to a new test. I can work on it.

adrianduffell commented on src/API/ProductsLowInStock.php:
```diff
@@ -0,0 +1,309 @@
+		$use_sitewide_stock_threshold = $this->is_using_sitewide_stock_threshold_only();
+
+		if ( $use_sitewide_stock_threshold ) {
+			$query_string = $this->get_query( false );
```
Should this be `true`?
```suggestion
			$query_string = $this->get_query( true );
```

Might be easier to read refactoring it like:

```
$query_string = $this->get_query( $use_sitewide_stock_threshold );
```

moon0326 replied:
Thank you! I'll make an update.

moon0326 replied:
Updated in https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/43e31e419c6798d04079c5330ecd2ae428f6deeb

Commit:
Use wc_rest_prepare_date_response to format the response date

adrianduffell commented on src/API/ProductsLowInStock.php:
```diff
@@ -0,0 +1,309 @@
+	 *
+	 * @return bool
+	 */
+	public function is_using_sitewide_stock_threshold_only() {
```
Nitpick: Could this be declared a `private` or `protected` method?

Further nitpick: you could switch all of the `protected` methods to `private` to help avoid backwards incompatible issues with inheritance. More info https://developer.woocommerce.com/2020/10/30/developer-advisory-inheritance-in-php-8/

moon0326 replied:
Good catch 👍 updated in https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/426ecb0a5405298cee0c469e620e1337b2d7b3d8

I think the problem described in the article is not related. It talks about when someone extends this new class and uses a mismatching method signature. I don't expect anyone to extend this class, but they should check the error in case they decide to extend this class.

adrianduffell replied:
> I don't expect anyone to extend this class, but they should check the error in case they decide to extend this class

The problem is if a third party has extended the class, we can’t then go and change the method signature without breaking things for them because we will create the mismatching signature. Hope that makes sense.

> Converting these methods to private prevent extending this class. If that's the case, we can simply make the class final

I think we should prevent extending the class through inheritance unless there is a use-case for it. Good idea about `final`, I'm happy with that 🙂

moon0326 replied:
@adrianduffell Ah! I see what you mean now. Thank you for the explanation. Yup, `final` sounds good. I'll make the change.

moon0326 replied:
updated in https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/5865208985af1284606af63c221dfc5ec22d9143

Commit:
Extract low in stock test to a separate file

Commit:
Make is_using_sitewide_stock_threshold_only protected

moon0326 commented:
> Awesome, thanks @moon0326
> 
> It looks like the cache header broke a test. I'm investigating.
> 
> On a related note, we probably need new PHP tests for the endpoint. In theory it could be a copy/paste of what's there for the deprecated endpoint?

Fixed the test and extracted to a new test file in https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/858c10b35a5473ccb8336e0ae5bacfed6809dd55

adrianduffell commented on src/API/ProductsLowInStock.php:
```diff
@@ -0,0 +1,309 @@
+		$low_stock_threshold          = absint( max( get_option( 'woocommerce_notify_low_stock_amount' ), 1 ) );
+		$use_sitewide_stock_threshold = $this->is_using_sitewide_stock_threshold_only();
+
+		if ( $use_sitewide_stock_threshold ) {
+			$query_string = $this->get_query( $use_sitewide_stock_threshold );
+		} else {
+			$query_string = $this->get_query();
+		}
```
Ah sorry, before I meant refactoring the whole `if-else` block like:
```suggestion
		$query_string = $this->get_query( $use_sitewide_stock_threshold );
```

I'm happy if you want to keep the `if-else` here. In that case it’s perhaps just better to pass `true` / `false` to `get_query()` like you originally had 😅

moon0326 replied:
I can't keep `if-else` after seeing this comment. haha :) I'll make the change.

moon0326 replied:
updated in https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/ca14814b2c87417b44dae95e516d3600b66dfa28

adrianduffell commented on src/API/Products.php:
```diff
@@ -112,12 +111,14 @@ protected function prepare_objects_query( $request ) {
 	public function get_items( $request ) {
-		add_filter( 'posts_fields', array( __CLASS__, 'add_wp_query_fields' ), 10, 2 );
+		if ( true === $request->get_param( 'low_in_stock' ) ) {
+			return ( new ProductsLowInStock() )->get_items( $request );
+		}
+
```
Sorry @moon0326, I think I've had a change of mind about this after testing today.

I think it would be better to keep the old endpoint exactly as it was for the time being. At the moment we don't know what else is using the old endpoint. I just worry there might be some combination of parameters being passed to it that we haven't replicated in the new endpoint yet. I'd hate to mess up a use-case in the mobile app for example.

I think I'd like to learn about this further before we make changes to the old endpoint. We could do it within a major release where there is more time to make sure we've got everything right.

moon0326 replied:
Sounds good to me. I'll revert the changes (controller + test) 👍

moon0326 replied:
updated in https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/b0a21fe087746e4317bce2c0c9b23e3f1cb9aa61 https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/8829d3426d8eaba682b9d8abd299b0bf31f58d8f

adrianduffell replied:
Much thanks !

Commit:
Revert Products.php changes

Commit:
Add back low in stock test to products test

Commit:
Remove unnecessary if

Commit:
Make the controller final

Commit:
Deprecate low_in_stock

Commit:
Bug fix -- last_order_date was not setting correctly for variation post type. This commit fixes it

adrianduffell reviewed (approved):
@moon0326 This is super work, thank you 🎉 👍 

It is testing well for me. Thanks for working on my change requests!

Let's merge this 🚀

Commit:
Add changelog
